cocos/core/data/class.ts

```diff
@@ -141,36 +141,36 @@ function getDefault (defaultVal): any {
     return defaultVal;
 }
 
-function doDefine (className, baseClass, options): any {
-    const ctor = options.ctor;
+function doDefine (className: string | undefined, baseClass: Constructor | null, options: Parameters<typeof CCClass>[0]): Constructor {
+    const ctor = options.ctor as Constructor;
 
     if (DEV) {
         // check ctor
         if (CCClass._isCCClass(ctor)) {
-            errorID(3618, className);
+            errorID(3618, className!);
         }
     }
 
     js.value(ctor, CCCLASS_TAG, true, true);
 
-    const prototype = ctor.prototype;
     if (baseClass) {
-        ctor.$super = baseClass;
+        // TODO: this is a dynamic inject method, should be define in class
+        // issue: https://github.com/cocos/cocos-engine/issues/14643
+        (ctor as any).$super = baseClass;
     }
 
-    js.setClassName(className, ctor);
+    js.setClassName(className!, ctor);
     return ctor;
 }
 
-function define (className, baseClass, options): any {
+function define (className: string | undefined, baseClass: Constructor | null, options: Parameters<typeof CCClass>[0]): Constructor {
```

Comment:
Why class name can be `undefined`?

Reply:
It's not a good interface design, but as it is used in `CCClass` method, it should be typeof `string | undefined`

https://github.com/cocos/cocos-engine/blob/c9d2f14086f45b18e0eccafdaf8b93d1b4200017/cocos/core/data/class.ts#L286-L297

Reply:
Can it work if passing undefined?

Reply:
yes, actually in the past, `undefined` was often passed in

Reply:
I think this line fix the `undefined` issue
https://github.com/cocos/cocos-engine/blob/c9d2f14086f45b18e0eccafdaf8b93d1b4200017/cocos/core/data/class.ts#L178